### PR TITLE
Used reserved resources firstly

### PR DESCRIPTION
**Pull request by k82cn**
fixes [MYRIAD-235](https://issues.apache.org/jira/browse/MYRIAD-235)

### CODE REVIEW

**Comment by DarinJ** [comment]
@adam-mesos can you take a 👀 .  I'm not super up on reserved resources.

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java [hunk 1]
```diff
@@ -81,83 +83,115 @@ public void onEvent(ResourceOffersEvent event, long sequence, boolean endOfBatch
...
+
+    for (Offer offer : offers) {
+      if (isReserved(offer)) {
+        reserved.add(offer);
+      } else {
+        normals.add(offer);
+      }
+    }
+
+    LOGGER.debug("Received total offers {}, including reserved offers {}", offers.size(), reserved.size());
```
Comment (yufeldman) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java]: I would use "reserved resources", not "reserved offers" phrase.

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java [hunk 2]
```diff
@@ -81,83 +83,115 @@ public void onEvent(ResourceOffersEvent event, long sequence, boolean endOfBatch
       }
       return;
     }
-    LOGGER.debug("Received offers {}", offers.size());
+
+    List<Offer> reserved = new ArrayList<>();
+    List<Offer> normals = new ArrayList<>();
```
Comment (yufeldman) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java]: I would change to "regular"

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java [hunk 3]
```diff
@@ -81,83 +83,115 @@ public void onEvent(ResourceOffersEvent event, long sequence, boolean endOfBatch
...
-                taskToLaunch.setHostname(offer.getHostname());
-                taskToLaunch.setSlaveId(offer.getSlaveId());
-                schedulerState.addTask(pendingTaskId, taskToLaunch);
-                iterator.remove(); // remove the used offer from offers list
-                break;
-              } catch (Throwable t) {
-                LOGGER.error("Exception thrown while trying to create a task for {}", taskPrefix, t);
-              }
-            }
+      dispatchTasks(driver, reserved);
```
Comment (yufeldman) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java]: You could concatenate the list (unreserved add after reserved) and call dispatchTasks just once

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java [hunk 4]
```diff
@@ -81,83 +83,115 @@ public void onEvent(ResourceOffersEvent event, long sequence, boolean endOfBatch
       }
       return;
     }
-    LOGGER.debug("Received offers {}", offers.size());
+
+    List<Offer> reserved = new ArrayList<>();
+    List<Offer> normals = new ArrayList<>();
+
+    for (Offer offer : offers) {
```
Comment (yufeldman) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java]: You could get fancy here and use sort where comparator does what your isReserved() method is doing.

**Comment by k82cn** [comment]
@yufeldman , any more comments?

**Comment by yufeldman** [comment]
Looks good - I did not realize you made changes, as it just showed close and reopen PR. Could you next time update PR, so it will be more obvious?

**Comment by k82cn** [comment]
Thanks for your inputs :).
Close by mistake, sorry.

**Comment by k82cn** [comment]
@DarinJ , @adam-mesos , any more comments? For the principal part, I'm going to open a new PR.

**Comment by yufeldman** [comment]
Sorry @k82cn - I can not checkin - looks like not enough privileges yet.
@adam-mesos, @DarinJ - could you merge this PR, unless you have more comments?

**Comment by DarinJ** [comment]
@k82cn give us a day please, I'd like to see if we can get @yufeldman access to merge.  If not I'll do it tomorrow.

**Comment by yufeldman** [comment]
@k82cn - thank you for your patience. PR has been merged.

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java [hunk 5]
```diff
@@ -83,6 +85,31 @@ public void onEvent(ResourceOffersEvent event, long sequence, boolean endOfBatch
     }
     LOGGER.debug("Received offers {}", offers.size());
     LOGGER.debug("Pending tasks: {}", this.schedulerState.getPendingTaskIds());
+
+    // Let Myriad use reserved resources firstly.
+    Collections.sort(offers, new Comparator<Offer>() {
+      boolean isReserved(Offer o) {
+        for (Protos.Resource resource : o.getResourcesList()) {
+          if (resource.hasRole() && !Objects.equals(resource.getRole(), DEFAULT_ROLE)) {
```
Comment (adam-mesos) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java]: You shouldn't be comparing against the default role if we allow users to specify a custom role. If a user specifies a non-default role, this code will ignore all resources reserved for it.
But Mesos will never offer resources for a role your framework is not registered for, so checking the value of the role is unnecessary anyway. Just bubbling up any offers with a role set is sufficient for this.

**Comment by adam-mesos** [comment]
Just a little bit slow. :stuck_out_tongue_closed_eyes:

**Pull request merged**